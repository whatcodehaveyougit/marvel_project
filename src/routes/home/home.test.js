import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import * as reactRedux from 'react-redux'
import Home from './home';
import configureStore from 'redux-mock-store';
import characterData from './testData/character-object.json';
jest.mock('../../components/spinner/spinner.component', () => jest.fn(() => '<Spinner />'));
jest.mock('../../components/character-card/character-card.component', () => jest.fn(() => '<CharacterCard />'));
// jest.mock('../../components/input/input.component', () => jest.fn(() => '<CustomInput />'));

const mockStore = configureStore([]);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'), // Preserve all original exports
  useSelector: jest.fn(),
}));

describe('Home', () => {
  it('renders correctly with character data', () => {
    const initialState = {
      characters: {
        characters: [ characterData ],
        isLoading: false,
        error: null,
      },
      character: {
        comics: [],
        data: null,
        isLoading: true,
        error: null
      }
    };
    const store = mockStore(initialState);
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(callback => callback(store.getState()));;
    const tree = render(<Provider store={store}><Home /></Provider>).asFragment();
    expect(tree).toMatchSnapshot();
  });
  it('renders spinner when data is loading', () => {
    const initialState = {
      characters: {
        characters: [  ],
        isLoading: true,
        error: null,
      },
      character: {
        comics: [],
        data: null,
        isLoading: true,
        error: null
      }
    };
    const store = mockStore(initialState);
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(callback => callback(store.getState()));;
    const tree = render(<Provider store={store}><Home /></Provider>).asFragment();
    expect(tree).toMatchSnapshot();
  });
  it('renders error correctly', () => {
    const initialState = {
      characters: {
        characters: [  ],
        isLoading: false,
        error: 'This is an ERROR!',
      },
      character: {
        comics: [],
        data: null,
        isLoading: true,
        error: null
      }
    };
    const store = mockStore(initialState);
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(callback => callback(store.getState()));;
    const tree = render(<Provider store={store}><Home /></Provider>).asFragment();
    expect(tree).toMatchSnapshot();
  });
  it('user types into input', () => {
    console.log(screen)
    render(<Home />);

    const input = screen.queryByTestId('custom-input')
    fireEvent.change(input, { target: { value: '123' } })
    expect(input).toHaveValue("123")
  })
});
