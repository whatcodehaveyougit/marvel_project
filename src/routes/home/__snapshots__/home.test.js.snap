// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly with character data 1`] = `
<DocumentFragment>
  <div
    class="home-container"
  >
    <div
      class="search-bar-component"
    >
      <input
        class="border border-gray-300 rounded-md p-2"
        data-testid="custom-input"
        name="search-monsters"
        placeholder="Search Marvel Characters..."
        type="text"
        value=""
      />
    </div>
    <div
      class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 gap-4"
    >
      <div
        class="character-card-container"
      >
        &lt;CharacterCard /&gt;
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home renders error correctly 1`] = `
<DocumentFragment>
  <div
    class="home-container"
  >
    <div
      class="search-bar-component"
    >
      <input
        class="border border-gray-300 rounded-md p-2"
        data-testid="custom-input"
        name="search-monsters"
        placeholder="Search Marvel Characters..."
        type="text"
        value=""
      />
    </div>
    <div
      class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 gap-4"
    />
    <div
      class="text-center text-red-700 text-2xl"
    >
      There was an error fetching the data
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home renders spinner when data is loading 1`] = `
<DocumentFragment>
  <div
    class="home-container"
  >
    <div
      class="search-bar-component"
    >
      <input
        class="border border-gray-300 rounded-md p-2"
        data-testid="custom-input"
        name="search-monsters"
        placeholder="Search Marvel Characters..."
        type="text"
        value=""
      />
    </div>
    <div>
      &lt;Spinner /&gt;
    </div>
  </div>
</DocumentFragment>
`;
